lib/bridge/PublicitySyncer.js

```diff
@@ -0,0 +1,157 @@
+}
+
+/* Solve any inconsistencies between the currently known state of channels '+s' modes
+   and rooms 'visibility' states. This does full graph traversal to prevent any +s
+   channels ever escaping into a 'public' room. This function errs on the side of
+   caution by assuming an unknown channel state is '+s'. This just means that if the
+   modes of a channel are not received yet (e.g when no virtual user is in said channel)
+   then the room is assumed secret (+s).
+
+   The bare minimum is done to make sure no private channels are leaked into public
+   matrix channels. If ANY +s channel is somehow being bridged into a room, that room
+   is updated to private. If ALL channels somehow being bridged into a room, that
+   room is allowed to be public.
+*/
+PublicitySyncer.prototype._solveVisibility = Promise.coroutine(function*() {
+    // For each room, do a big OR on all of the channels that are linked in any way
+
+    this._visibilityMap.mappings = yield this.ircBridge.getStore().getAllChannelMappings();
+
+    let roomIds = Object.keys(this._visibilityMap.mappings);
+
+    let shouldBePrivate = (roomId, checkedChannels) => {
+        // If any channel connected to this room is +s, stop early and call it private
+
+        // List first connected
+        let channels = this._visibilityMap.mappings[roomId];
+        //      = ['localhost #channel1', 'localhost #channel2', ... ]
+
+        // No channels mapped to this roomId
+        if (!channels) {
+            return false;
+        }
+
+        // Filter out already checked channels
+        channels = channels.filter((c) => checkedChannels.indexOf(c) === -1);
+
+        let anyAreSecret = channels.some((channel) => {
+            let channelIsSecret = this._visibilityMap.channelIsSecret[channel];
+
+            // If a channel mode is unknown, assume it is secret
+            if (typeof channelIsSecret === 'undefined') {
+                log.info('Assuming channel ' + channel + ' is secret');
+                channelIsSecret = true;
+            }
+
+            return channelIsSecret;
+        });
+        if (anyAreSecret) {
+            return true;
+        }
+
+        // Otherwise, recurse with the rooms connected to each channel
+
+        // So get all the roomIds that this channel is mapped to and return whether any
+        //  are mapped to channels that are secret
+        return channels.map((channel) => {
+            return Object.keys(this._visibilityMap.mappings).filter((roomId2) => {
+                return this._visibilityMap.mappings[roomId2].indexOf(channel) !== -1;
+            });
+        }).some((roomIds2) => {
+            return roomIds2.some((roomId2) => {
+                return shouldBePrivate(roomId2, checkedChannels.concat(channels));
+            });
+        });
+    }
+
+    let roomCorrectVisibilities = roomIds.map(
+        (roomId) => {
+            return shouldBePrivate(roomId, []);
+        }
+    ).map((b) => b ? 'private' : 'public');
```

Comment:
I still don't like how we index into this array on `:142`. Can we not just put:
```js
let correctState = shouldBePrivate(roomId, []) ? "private": "public";
```
on `:142`? One less loop and it's clearer.